### PR TITLE
U4-8960 Misc classes have hard dependencies on HttpContext.Server.MapPath

### PR DESCRIPTION
**Pull request by lars-erik**
Removing HttpContext dependency from GridValueConverter. Adding configuration for IOHelper when testing to remove usage of HttpContext.

Don't pull it immediately. Working on a presentation where I test drive views. Not quite done yet, so might stumble upon some more showstoppers.

### CODE REVIEW

File: src/Umbraco.Core/IO/IOHelper.cs [hunk 1]
```diff
@@ -13,7 +13,9 @@
 namespace Umbraco.Core.IO
 {
 	public static class IOHelper
-    {
+	{
+	    public static bool IAmUnitTestingSoNeverUseHttpContextEver = false;
```
Comment (Shazwazza) [inline src/Umbraco.Core/IO/IOHelper.cs]: lol! ... ok fair enough :) Out of curiosity how come the null check on HttpContext.Current doesn't work for your test scenario and you need this flag?

Comment (lars-erik) [inline src/Umbraco.Core/IO/IOHelper.cs]: :)
HttpContext.Current isn't null, but you can't use Server.MapPath and HostingEnvironment.MapPath when unit testing.

Comment (lars-erik) [inline src/Umbraco.Core/IO/IOHelper.cs]: See https://github.com/lars-erik/specflow-razorgenerator-umbraco-samples for context.

Comment (Shazwazza) [inline src/Umbraco.Core/IO/IOHelper.cs]: ah of course! I'd like to kill this class, can't remember what has been done in v8 so far, i think it's an injected service but haven't looked at that for a while.

Comment (lars-erik) [inline src/Umbraco.Core/IO/IOHelper.cs]: Yep, injected will be way to go. Wish there was HostingEnvironment.Set(Mode.UnitTest, path).

File: src/Umbraco.Tests/TestHelpers/BaseDatabaseFactoryTest.cs [hunk 2]
```diff
@@ -375,13 +375,12 @@ protected UmbracoContext GetUmbracoContext(string url, int templateId, RouteData
             return ctx;
         }
 
-        protected FakeHttpContextFactory GetHttpContextFactory(string url, RouteData routeData = null)
+        protected virtual FakeHttpContextFactory GetHttpContextFactory(string url, RouteData routeData = null)
```
Comment (lars-erik) [inline src/Umbraco.Tests/TestHelpers/BaseDatabaseFactoryTest.cs]: This might actually be YAGNI. I'll give it another shot this evening.

**Comment by zpqrtbnk** [comment]
>Don't pull it immediately. 
Hey... been a year ;-) What's the status of this PR? Should we just kill it, or?

**Comment by lars-erik** [comment]
I haven't hit anything else, but I don't think I've hit the issue any more either.
It can't hurt to pull, tho.

**Comment by zpqrtbnk** [comment]
k so it's not "work in progress" anymore really, right? will look into it then.

**Comment by lars-erik** [comment]
No WIP.

There's the blatant dependency on the IIS pipeline / AppDomain (MapPath), and the other one is just nice OCP.

**Comment by zpqrtbnk** [comment]
would it be ok to have one single `IAmUnitTestingBlah` property on `IOHelper` that would also be used by `TypeFinder` or is it important that they are two different properties?

**Comment by zpqrtbnk** [comment]
have fixed-merge, hopefully it's ok for you else let me know

**Comment by lars-erik** [comment]
Although, it should preferably show that it's just for unit testing. To discourage people from making weird assumptions in production code. Or maybe get a nice xml summary saying that it will break in web contexts.

**Pull request merged**